disttask/framework/scheduler/manager_test.go

```diff
@@ -0,0 +1,209 @@
+		return mockInternalScheduler
+	}
+	originPoolFunc := newPool
+	newPool = func(name string, size int32, component util.Component, options ...spool.Option) (Pool, error) {
+		return mockPool, nil
+	}
+	RegisterSubtaskExectorConstructor("type", func(minimalTask proto.MinimalTask, step int64) (SubtaskExecutor, error) {
+		return &MockSubtaskExecutor{}, nil
+	}, func(opts *subtaskExecutorRegisterOptions) {
+		opts.PoolSize = 1
+	})
+	defer func() {
+		newScheduler = originSchedulerFunc
+		newPool = originPoolFunc
+		delete(subtaskExecutorConstructors, "type")
+		delete(subtaskExecutorOptions, "type")
+	}()
+	id := "test"
+	taskID1 := int64(1)
+	taskID2 := int64(2)
+	task1 := &proto.Task{ID: taskID1, State: proto.TaskStateRunning, Step: 0, Type: "type"}
+	task2 := &proto.Task{ID: taskID2, State: proto.TaskStateReverting, Step: 0, Type: "type"}
+
+	mockTaskTable.On("GetTasksInStates", proto.TaskStateRunning, proto.TaskStateReverting).Return([]*proto.Task{task1, task2}, nil)
+	mockTaskTable.On("GetTasksInStates", proto.TaskStateReverting).Return([]*proto.Task{task2}, nil)
+	// task1
+	mockSubtaskTable.On("HasSubtasksInStates", id, taskID1, []interface{}{proto.TaskStatePending, proto.TaskStateRevertPending}).Return(true, nil).Once()
+	mockPool.On("Run", mock.Anything).Return(nil).Once()
+	mockInternalScheduler.On("Start").Once()
+	mockTaskTable.On("GetTaskByID", taskID1).Return(task1, nil)
+	mockSubtaskTable.On("HasSubtasksInStates", id, taskID1, []interface{}{proto.TaskStatePending, proto.TaskStateRevertPending}).Return(true, nil).Once()
+	mockInternalScheduler.On("Run", mock.Anything, task1).Return(nil).Once()
+	mockSubtaskTable.On("HasSubtasksInStates", id, taskID1, []interface{}{proto.TaskStatePending, proto.TaskStateRevertPending}).Return(false, nil)
+	mockInternalScheduler.On("Stop").Once()
+	// task2
+	mockSubtaskTable.On("HasSubtasksInStates", id, taskID2, []interface{}{proto.TaskStatePending, proto.TaskStateRevertPending}).Return(true, nil).Once()
+	mockPool.On("Run", mock.Anything).Return(nil).Once()
+	mockInternalScheduler.On("Start").Once()
+	mockTaskTable.On("GetTaskByID", taskID2).Return(task2, nil)
+	mockSubtaskTable.On("HasSubtasksInStates", id, taskID2, []interface{}{proto.TaskStatePending, proto.TaskStateRevertPending}).Return(true, nil).Once()
+	mockInternalScheduler.On("Rollback", mock.Anything, task2).Return(nil).Once()
+	mockSubtaskTable.On("HasSubtasksInStates", id, taskID2, []interface{}{proto.TaskStatePending, proto.TaskStateRevertPending}).Return(false, nil)
+	mockInternalScheduler.On("Stop").Once()
+	mockPool.On("ReleaseAndWait").Twice()
+	RegisterSubtaskExectorConstructor("type", func(minimalTask proto.MinimalTask, step int64) (SubtaskExecutor, error) {
+		return &MockSubtaskExecutor{}, nil
+	}, func(opts *subtaskExecutorRegisterOptions) {
+		opts.PoolSize = 1
+	})
+	m, err := NewManager(context.Background(), id, mockTaskTable, mockSubtaskTable)
```

Comment:
Could we use `storage.NewGlobalTaskManager` instead of `mockTaskTable`?
I think it would be better to just call real code if it weren't for some problems that aren't very well simulated.

Reply:
I think using mock can be a better way to test the behavior of the manager, such as the times and parameters of subtaskTaskTable calls. If we want to test the behavior of all, it is better to use unit test of example or integration test

Reply:
The main feeling is that the operations of the dispatcher can be mock (out of the scheduler's control), while the scheduler's internal processing of the subtaskTable can be done without mock.
However, the tests here are comprehensive and mock calls to failed (previously handled with failpoint), so it makes sense to introduce mocks.